### PR TITLE
lsp: Implement first pass at package and rule ref based completions

### PR DESCRIPTION
**Pull request by charlieegan3**
Fixes https://github.com/StyraInc/regal/issues/724

https://github.com/StyraInc/regal/assets/1774239/40f630bc-03d5-4f00-acb3-c6530d1943da

### CODE REVIEW

File: internal/lsp/completions/providers/packagerefs.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import (
+	"strings"
+
+	"github.com/styrainc/regal/internal/lsp/cache"
+	"github.com/styrainc/regal/internal/lsp/types"
+	"github.com/styrainc/regal/internal/lsp/types/symbols"
+)
+
+// PackageRefs is a completion provider that returns completions when importing packages.
+type PackageRefs struct{}
```
Comment (charlieegan3) [inline internal/lsp/completions/providers/packagerefs.go:12]: This is one of the new ref based completion providers added in this PR

File: internal/lsp/completions/providers/regov1.go [hunk 2]
```diff
@@ -12,29 +12,21 @@ type RegoV1 struct{}
 func (*RegoV1) Run(c *cache.Cache, params types.CompletionParams, _ *Options) ([]types.CompletionItem, error) {
 	fileURI := params.TextDocument.URI
 
-	fileContents, ok := c.GetFileContents(fileURI)
-	if !ok {
-		// if the file contents is missing then we can't provide completions
-		return nil, nil
+	lines, currentLine := completionLineHelper(c, fileURI, params.Position.Line)
```
Comment (charlieegan3) [inline internal/lsp/completions/providers/regov1.go:15]: This is refactoring  completionLineHelper shares some code needed in most completion providers.

File: internal/lsp/completions/providers/rulerefs.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	"strings"
+
+	"github.com/styrainc/regal/internal/lsp/cache"
+	"github.com/styrainc/regal/internal/lsp/types"
+	"github.com/styrainc/regal/internal/lsp/types/symbols"
+)
+
+// RuleFromImportedPackageRefs is a completion provider that returns completions for
+// rules found in already imported packages.
+type RuleFromImportedPackageRefs struct{}
```
Comment (charlieegan3) [inline internal/lsp/completions/providers/rulerefs.go:13]: This is a new ref based completion provider added in this PR

File: internal/lsp/completions/refs/refs.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
+package refs
```
Comment (charlieegan3) [inline internal/lsp/completions/refs/refs.go:1]: This package adds the ForModule function that backs all the new ref based completions.

File: internal/lsp/completions/refs/refs.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+	"github.com/open-policy-agent/opa/ast"
+
+	rast "github.com/styrainc/regal/internal/ast"
+	"github.com/styrainc/regal/internal/lsp/types"
+)
+
+// ForModule returns a map of refs and details about them to be used in completions that
+// were found in the given module.
+func ForModule(module *ast.Module) map[string]types.Ref {
```
Comment (charlieegan3) [inline internal/lsp/completions/refs/refs.go:18]: The refs returned are for rules and the package itself. Both are used in different providers.

File: internal/lsp/documentsymbol.go [hunk 6]
```diff
@@ -149,122 +149,6 @@ func locationToRange(location *ast.Location) types.Range {
 	}
 }
 
-func refToString(ref ast.Ref) string {
```
Comment (charlieegan3) [inline internal/lsp/documentsymbol.go:152]: These functions have been moved to the new Regal ast package, `rast`

File: internal/lsp/server.go [hunk 7]
```diff
@@ -680,17 +680,26 @@ func (l *LanguageServer) handleTextDocumentCompletion(
...
 		return nil, fmt.Errorf("failed to unmarshal params: %w", err)
 	}
 
+	// items is allocated here so that the return value is always a non-nil CompletionList
 	items, err := l.completionsManager.Run(params, &providers.Options{RootURI: l.clientRootURI})
 	if err != nil {
 		return nil, fmt.Errorf("failed to find completions: %w", err)
 	}
 
+	if len(items) == 0 {
```
Comment (charlieegan3) [inline internal/lsp/server.go:695]: not strictly related to this PR, but this makes sure that the items is never null.

**Review by anderseknert — COMMENTED** [review]
This is awesome, and the annotations processing was really something else! I'll take this for a test drive, but the code looks good to me 👍

File: internal/ast/ref.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package ast
+
+import (
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+)
+
+// RefToString converts an ast.Ref to a readable string, e.g. data.foo[bar].
+func RefToString(ref ast.Ref) string {
```
[the remark this answers is not in the document]
Comment (charlieegan3) [reply internal/ast/ref.go:10]: yeah not sure what'd happen there... but for now it works and the use cases for this area all tested so I think we're good.

[the remark this answers is not in the document]
Comment (anderseknert) [reply internal/ast/ref.go:10]: Fixing it would be pretty simple I think, but we can do that post merge.

**Review by anderseknert — APPROVED** [review]
Phenomenal! It's long enough to the next release, and I think this is big enough that I prefer to have this merged and do the testing on main, and create new issues for any follow-up concerns we might have.

Outstanding work! 👏

**Pull request merged**